### PR TITLE
add parallel actor critic algorithm

### PR DESCRIPTION
**Pull request by loofahcus**
There is something wrong with symbol definition and gradients calculation in [A3C](https://github.com/dmlc/mxnet/tree/master/example/reinforcement-learning/a3c),  here is the parallel actor critic algorithm transplanted from [Minpy#parallel_actor_critic](https://github.com/dmlc/minpy/tree/master/examples/rl/parallel_actor_critic). This implementation can get relatively good results for Pong in about 30 minutes of training.

### CODE REVIEW

**Comment by piiswrong** [comment]
Could you be more specific on whats wrong with the A3C implementation?

**Comment by loofahcus** [comment]
Hi @piiswrong, 
1.  [A3C](https://github.com/dmlc/mxnet/tree/master/example/reinforcement-learning/a3c) is actually parallel actor critic, since it uses batch learning instead of asynchronous learning.

2. As noeted in "Policy Gradient Theorem", in policy gradient, the loss is: , which just considers the action taken, so the policy output should be SoftmaxActivation(with advantage as out_gradient) instead of SoftmaxOutput.

3. I think the entropy loss in [A3C](https://github.com/dmlc/mxnet/tree/master/example/reinforcement-learning/a3c) is not right, please refer to [parallel_actor_critic/model.py#L30](https://github.com/loofahcus/mxnet/blob/parallel_actor_critic/example/reinforcement-learning/parallel_actor_critic/model.py#L30)

**Comment by sxjscience** [comment]
@loofahcus It's really great if we can get the A3C implementation correct! Have you tried it on other games like BreakOut? Also, I think that SoftmaxActivation with advantage as the out_grad should be the same as SoftmaxOutput with the advantage as the label. And A3C in the original paper should be asynchronous, which uses different exploring ratio for the agents.

**Comment by loofahcus** [comment]
@sxjscience Sorry, I didn't make it clear. When I said **"so the policy output should be SoftmaxActivation(with advantage as out_gradient) instead of SoftmaxOutput."**, I actually meant log(SoftmaxActivation), you can see [here](https://github.com/loofahcus/mxnet/blob/parallel_actor_critic/example/reinforcement-learning/parallel_actor_critic/model.py#L27), that's why I said SoftmaxOutput is not proper.

I myself didn't try other games like BreakOut, but I refered to other A3C implementation like [async-rl](https://github.com/muupan/async-rl) and quit sure the implementation is right. The [shared net](https://github.com/loofahcus/mxnet/blob/parallel_actor_critic/example/reinforcement-learning/parallel_actor_critic/model.py#L18) uses MLP instead of CNN, since [Atari8080Preprocessor](https://github.com/loofahcus/mxnet/blob/parallel_actor_critic/example/reinforcement-learning/parallel_actor_critic/envs.py#L4) simplifies the observation into “binary image”, which leads to fast learning, and so it may not fit other games. Since this implementation focuses on right implementation of network structure of actor-critic, it's not a big deal.

By the way, in my opinion, the biggest difference between **A3C** and **Parallel Actor Critic** may lie in the rate of convergence, A3C may benefit from its hogwild learning schema.

**Comment by sxjscience** [comment]
I think it's a good example for parallel actor critic.

**Comment by tonysy** [comment]
Hi, I have run this algorithm on my server(28  Intel(R) Xeon(R) CPU E5-2690 v4 @ 2.60GHz), with the default parameter setting(num-envs = 16). It takes too much time to solve this problem, more than 10 hours.  I use the latest version of mxnet. I would like to know what cause this algorithm to be solved so slowly. Thansk.
```
Batch 1061 complete (30.74s) (31801.8s elapsed) (episode 16992), batch avg. reward: 20.31, running reward:[938/1835]
Batch 1062 complete (32.07s) (31833.8s elapsed) (episode 17008), batch avg. reward: 17.00, running reward: 16.872
Batch 1063 complete (29.40s) (31863.2s elapsed) (episode 17024), batch avg. reward: 18.19, running reward: 17.056
Batch 1064 complete (31.82s) (31895.0s elapsed) (episode 17040), batch avg. reward: 14.12, running reward: 16.611
Batch 1065 complete (28.39s) (31923.4s elapsed) (episode 17056), batch avg. reward: 18.31, running reward: 16.843
Batch 1066 complete (28.56s) (31952.0s elapsed) (episode 17072), batch avg. reward: 18.31, running reward: 17.088
Batch 1067 complete (32.69s) (31984.7s elapsed) (episode 17088), batch avg. reward: 18.50, running reward: 17.277
Batch 1068 complete (33.23s) (32017.9s elapsed) (episode 17104), batch avg. reward: 19.44, running reward: 17.599
Batch 1069 complete (29.90s) (32047.8s elapsed) (episode 17120), batch avg. reward: 18.62, running reward: 17.774
Batch 1070 complete (28.10s) (32075.9s elapsed) (episode 17136), batch avg. reward: 15.81, running reward: 17.524
Batch 1071 complete (28.70s) (32104.6s elapsed) (episode 17152), batch avg. reward: 16.06, running reward: 17.347
Batch 1072 complete (29.84s) (32134.5s elapsed) (episode 17168), batch avg. reward: 18.75, running reward: 17.563
Batch 1073 complete (28.55s) (32163.0s elapsed) (episode 17184), batch avg. reward: 18.25, running reward: 17.668
Batch 1074 complete (30.55s) (32193.6s elapsed) (episode 17200), batch avg. reward: 14.31, running reward: 17.162
Batch 1075 complete (29.53s) (32223.1s elapsed) (episode 17216), batch avg. reward: 20.94, running reward: 17.723
Batch 1076 complete (28.64s) (32251.7s elapsed) (episode 17232), batch avg. reward: 18.38, running reward: 17.826
Batch 1077 complete (28.24s) (32280.0s elapsed) (episode 17248), batch avg. reward: 15.69, running reward: 17.507
Batch 1078 complete (32.42s) (32312.4s elapsed) (episode 17264), batch avg. reward: 14.56, running reward: 17.091
Batch 1079 complete (28.42s) (32340.8s elapsed) (episode 17280), batch avg. reward: 18.38, running reward: 17.302
Batch 1080 complete (32.11s) (32372.9s elapsed) (episode 17296), batch avg. reward: 15.81, running reward: 17.074
Batch 1081 complete (29.21s) (32402.1s elapsed) (episode 17312), batch avg. reward: 17.94, running reward: 17.221
Batch 1082 complete (31.46s) (32433.6s elapsed) (episode 17328), batch avg. reward: 16.50, running reward: 17.131
Batch 1083 complete (29.41s) (32463.0s elapsed) (episode 17344), batch avg. reward: 20.75, running reward: 17.670
Batch 1084 complete (28.96s) (32492.0s elapsed) (episode 17360), batch avg. reward: 18.56, running reward: 17.790
Batch 1085 complete (29.06s) (32521.0s elapsed) (episode 17376), batch avg. reward: 18.62, running reward: 17.933
Batch 1086 complete (31.93s) (32553.0s elapsed) (episode 17392), batch avg. reward: 15.31, running reward: 17.507
Batch 1087 complete (31.80s) (32584.7s elapsed) (episode 17408), batch avg. reward: 20.12, running reward: 17.890
Batch 1088 complete (28.33s) (32613.1s elapsed) (episode 17424), batch avg. reward: 13.44, running reward: 17.215
Batch 1089 complete (31.49s) (32644.6s elapsed) (episode 17440), batch avg. reward: 5.94, running reward: 15.559
Batch 1090 complete (28.32s) (32672.9s elapsed) (episode 17456), batch avg. reward: 15.88, running reward: 15.631
Batch 1091 complete (29.21s) (32702.1s elapsed) (episode 17472), batch avg. reward: 18.56, running reward: 16.079
Batch 1092 complete (28.98s) (32731.1s elapsed) (episode 17488), batch avg. reward: 18.44, running reward: 16.443
Batch 1093 complete (29.88s) (32761.0s elapsed) (episode 17504), batch avg. reward: 18.06, running reward: 16.655
Batch 1094 complete (29.35s) (32790.3s elapsed) (episode 17520), batch avg. reward: 18.12, running reward: 16.868
Batch 1095 complete (29.36s) (32819.7s elapsed) (episode 17536), batch avg. reward: 12.88, running reward: 16.280
Batch 1096 complete (29.65s) (32849.3s elapsed) (episode 17552), batch avg. reward: 18.75, running reward: 16.636
Batch 1097 complete (31.78s) (32881.1s elapsed) (episode 17568), batch avg. reward: 19.62, running reward: 17.077
Batch 1098 complete (29.47s) (32910.6s elapsed) (episode 17584), batch avg. reward: 15.12, running reward: 16.774
Batch 1099 complete (30.07s) (32940.6s elapsed) (episode 17600), batch avg. reward: 20.88, running reward: 17.382
Batch 1100 complete (30.22s) (32970.9s elapsed) (episode 17616), batch avg. reward: 15.62, running reward: 17.124
Batch 1101 complete (33.70s) (33004.6s elapsed) (episode 17632), batch avg. reward: 17.31, running reward: 17.167
Batch 1102 complete (32.87s) (33037.4s elapsed) (episode 17648), batch avg. reward: 14.56, running reward: 16.810
Batch 1103 complete (29.89s) (33067.3s elapsed) (episode 17664), batch avg. reward: 18.62, running reward: 17.073
Batch 1104 complete (29.67s) (33097.0s elapsed) (episode 17680), batch avg. reward: 18.44, running reward: 17.298
Batch 1105 complete (30.19s) (33127.2s elapsed) (episode 17696), batch avg. reward: 16.06, running reward: 17.145
Batch 1106 complete (30.90s) (33158.1s elapsed) (episode 17712), batch avg. reward: 14.06, running reward: 16.678
Batch 1107 complete (34.82s) (33192.9s elapsed) (episode 17728), batch avg. reward: 17.19, running reward: 16.770
Batch 1108 complete (29.23s) (33222.1s elapsed) (episode 17744), batch avg. reward: 17.94, running reward: 16.962
Batch 1109 complete (28.70s) (33250.8s elapsed) (episode 17760), batch avg. reward: 18.25, running reward: 17.158
Batch 1110 complete (28.27s) (33279.1s elapsed) (episode 17776), batch avg. reward: 15.50, running reward: 16.960
Batch 1111 complete (29.08s) (33308.2s elapsed) (episode 17792), batch avg. reward: 13.62, running reward: 16.446
Batch 1112 complete (32.80s) (33341.0s elapsed) (episode 17808), batch avg. reward: 14.88, running reward: 16.227
Batch 1113 complete (32.26s) (33373.2s elapsed) (episode 17824), batch avg. reward: 17.19, running reward: 16.361
Batch 1114 complete (33.19s) (33406.4s elapsed) (episode 17840), batch avg. reward: 19.94, running reward: 16.898
Batch 1115 complete (31.45s) (33437.9s elapsed) (episode 17856), batch avg. reward: 16.62, running reward: 16.832
Batch 1116 complete (32.06s) (33469.9s elapsed) (episode 17872), batch avg. reward: 13.75, running reward: 16.413
Batch 1117 complete (34.21s) (33504.1s elapsed) (episode 17888), batch avg. reward: 17.62, running reward: 16.565
Batch 1118 complete (31.35s) (33535.5s elapsed) (episode 17904), batch avg. reward: 16.19, running reward: 16.543
Batch 1119 complete (29.52s) (33565.0s elapsed) (episode 17920), batch avg. reward: 20.75, running reward: 17.167
Batch 1120 complete (33.63s) (33598.7s elapsed) (episode 17936), batch avg. reward: 13.94, running reward: 16.722
Batch 1121 complete (30.43s) (33629.1s elapsed) (episode 17952), batch avg. reward: 15.81, running reward: 16.578
Batch 1122 complete (28.45s) (33657.5s elapsed) (episode 17968), batch avg. reward: 7.81, running reward: 15.319
Batch 1123 complete (29.82s) (33687.4s elapsed) (episode 17984), batch avg. reward: 18.00, running reward: 15.710
Batch 1124 complete (33.27s) (33720.6s elapsed) (episode 18000), batch avg. reward: 18.81, running reward: 16.157
Batch 1125 complete (28.94s) (33749.6s elapsed) (episode 18016), batch avg. reward: 15.50, running reward: 16.049
Batch 1126 complete (32.01s) (33781.6s elapsed) (episode 18032), batch avg. reward: 17.56, running reward: 16.271
Batch 1127 complete (33.79s) (33815.4s elapsed) (episode 18048), batch avg. reward: 16.75, running reward: 16.298
Batch 1128 complete (29.23s) (33844.6s elapsed) (episode 18064), batch avg. reward: 20.81, running reward: 16.969
Batch 1129 complete (34.03s) (33878.6s elapsed) (episode 18080), batch avg. reward: 12.38, running reward: 16.258
Batch 1130 complete (29.88s) (33908.5s elapsed) (episode 18096), batch avg. reward: 20.50, running reward: 16.888
Batch 1131 complete (27.83s) (33936.3s elapsed) (episode 18112), batch avg. reward: 10.19, running reward: 15.899
Batch 1132 complete (29.85s) (33966.2s elapsed) (episode 18128), batch avg. reward: 19.12, running reward: 16.382
Batch 1133 complete (28.03s) (33994.2s elapsed) (episode 18144), batch avg. reward: 15.69, running reward: 16.316
Batch 1134 complete (28.33s) (34022.5s elapsed) (episode 18160), batch avg. reward: 13.44, running reward: 15.900
Batch 1135 complete (32.37s) (34054.9s elapsed) (episode 18176), batch avg. reward: 19.00, running reward: 16.374
Batch 1136 complete (30.28s) (34085.2s elapsed) (episode 18192), batch avg. reward: 16.50, running reward: 16.376
Batch 1137 complete (29.31s) (34114.5s elapsed) (episode 18208), batch avg. reward: 20.81, running reward: 17.036
Batch 1138 complete (28.40s) (34142.9s elapsed) (episode 18224), batch avg. reward: 12.81, running reward: 16.432
Batch 1139 complete (29.99s) (34172.9s elapsed) (episode 18240), batch avg. reward: 13.88, running reward: 16.019
Batch 1140 complete (29.17s) (34202.0s elapsed) (episode 18256), batch avg. reward: 18.12, running reward: 16.307
Batch 1141 complete (28.81s) (34230.9s elapsed) (episode 18272), batch avg. reward: 18.06, running reward: 16.593
Batch 1142 complete (28.09s) (34258.9s elapsed) (episode 18288), batch avg. reward: 15.75, running reward: 16.500
........
.........
Batch 1722 complete (28.69s) (51759.5s elapsed) (episode 27568), batch avg. reward: 21.00, running reward: 18.450
Batch 1723 complete (30.06s) (51789.5s elapsed) (episode 27584), batch avg. reward: 17.88, running reward: 18.347
Batch 1724 complete (30.61s) (51820.1s elapsed) (episode 27600), batch avg. reward: 16.62, running reward: 18.099
Batch 1725 complete (29.28s) (51849.4s elapsed) (episode 27616), batch avg. reward: 20.88, running reward: 18.512
Batch 1726 complete (29.61s) (51879.0s elapsed) (episode 27632), batch avg. reward: 20.75, running reward: 18.845
Batch 1727 complete (28.95s) (51908.0s elapsed) (episode 27648), batch avg. reward: 18.44, running reward: 18.783
Batch 1728 complete (29.00s) (51937.0s elapsed) (episode 27664), batch avg. reward: 20.88, running reward: 19.095
Batch 1729 complete (29.87s) (51966.8s elapsed) (episode 27680), batch avg. reward: 20.62, running reward: 19.324
Batch 1730 complete (29.19s) (51996.0s elapsed) (episode 27696), batch avg. reward: 20.94, running reward: 19.564
Batch 1731 complete (30.47s) (52026.5s elapsed) (episode 27712), batch avg. reward: 20.62, running reward: 19.722
```
